src/org/jitsi/impl/neomedia/rtp/translator/SimulcastController.java

```diff
@@ -0,0 +1,533 @@
+     */
+    private int targetIdx = -1;
+
+    /**
+     * Read by the transform thread. Written by the filter thread.
+     */
+    private SimTransformation transformState = dropState;
+
+    /**
+     * Ctor.
+     *
+     * @param source the {@link MediaStreamTrackDesc} that feeds this instance
+     * with RTP/RTCP packets.
+     */
+    public SimulcastController(MediaStreamTrackDesc source)
+    {
+        this.weakSource = new WeakReference<>(source);
+        this.targetSSRC = source.getRTPEncodings()[0].getPrimarySSRC();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PacketTransformer getRTPTransformer()
+    {
+        return rtpTransformer;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PacketTransformer getRTCPTransformer()
+    {
+        return rtcpTransformer;
+    }
+
+    /**
+     *
+     * @param data
+     * @param buf
+     * @param off
+     * @param len
+     * @return
+     */
+    public boolean rtpTranslatorWillWrite(
+        boolean data, byte[] buf, int off, int len)
+    {
+        MediaStreamTrackDesc sourceTrack = weakSource.get();
+
+        // If we're getting packets here => the MST is alive.
+        assert sourceTrack != null;
+
+        long sourceSSRC;
+        if (data)
+        {
+            sourceSSRC = RawPacket.getSSRCAsLong(buf, off, len);
+        }
+        else
+        {
+            // NOTE(gp) This will need to be adjusted when we enable
+            // reduced-size RTCP.
+            sourceSSRC = RTCPHeaderUtils.getSenderSSRC(buf, off, len);
+            return sourceSSRC == transformState.currentSSRC;
+        }
+
+        boolean currentRTPEncodingIsActive = false;
+        if (transformState.currentIdx > -1)
+        {
+            RTPEncodingDesc
+                currentRTPEncodingDesc = sourceTrack
+                .getRTPEncodings()[transformState.currentIdx];
+
+            currentRTPEncodingIsActive = currentRTPEncodingDesc.isActive();
+        }
+
+        if (transformState.currentIdx == targetIdx && currentRTPEncodingIsActive)
+        {
+            boolean accept = sourceSSRC == transformState.currentSSRC;
+
+            if (accept)
+            {
+                onAccept(buf, off, len);
+            }
+
+            return accept;
+        }
+
+
+        // An intra-codec/simulcast switch pending.
+
+        FrameDesc sourceFrameDesc
+            = sourceTrack.findFrameDesc(buf, off, len);
+
+        int sourceIdx = sourceFrameDesc.getRTPEncoding().getIndex();
+
+        if (!sourceFrameDesc.isIndependent()
+            || sourceIdx == transformState.currentIdx)
+        {
+            // An intra-codec switch requires a key frame.
+            boolean accept = sourceSSRC == transformState.currentSSRC;
+
+            if (accept)
+            {
+                onAccept(buf, off, len);
+            }
+
+            return accept;
+        }
+
+        if (((targetIdx <= sourceIdx && sourceIdx < transformState.currentIdx)
```

Comment:
Why is this whole condition enclosed in two sets of brackets? I would expect that the conjunction on the third line be in brackets.